File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        # Ensure nums1 is the smaller array for efficient binary search
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
            
        m, n = len(nums1), len(nums2)  # Lengths of both arrays
        low, high = 0, m  # Set binary search bounds

        # Perform binary search
        while low <= high:
            # Partitioning index for nums1
            partitionX = (low + high) // 2
            # Corresponding partition index for nums2
            partitionY = (m + n + 1) // 2 - partitionX
            
            # Max and min values for left and right partitions
            maxX = float('-inf') if partitionX == 0 else nums1[partitionX - 1]
            maxY = float('-inf') if partitionY == 0 else nums2[partitionY - 1]
            minX = float('inf') if partitionX == m else nums1[partitionX]
            minY = float('inf') if partitionY == n else nums2[partitionY]
            
            # Check if we have found the correct partitions
            if maxX <= minY and maxY <= minX:
                # Even total length: return average of max left and min right
                if (m + n) % 2 == 0:
                    return (max(maxX, maxY) + min(minX, minY)) / 2
                # Odd total length: return max of left partitions
                else:
                    return max(maxX, maxY)
            # Move towards the left partition of nums1
            elif maxX > minY:
                high = partitionX - 1
            # Move towards the right partition of nums1
            else:
                low = partitionX + 1

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        // Ensure nums1 is the smaller array to optimize binary search
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.length; // Length of the first array
        int n = nums2.length; // Length of the second array
        int left = 0; // Start of the binary search on nums1
        int right = m; // End of the binary search on nums1

        // Perform binary search on the smaller array
        while (left <= right) {
            // Partition nums1 and nums2
            int partitionA = (left + right) / 2; // Current partition index for nums1
            int partitionB = (m + n + 1) / 2 - partitionA; // Corresponding partition index for nums2

            // Get max and min values around the partition for nums1
            int maxLeftA = (partitionA == 0) ? Integer.MIN_VALUE : nums1[partitionA - 1];
            int minRightA = (partitionA == m) ? Integer.MAX_VALUE : nums1[partitionA];

            // Get max and min values around the partition for nums2
            int maxLeftB = (partitionB == 0) ? Integer.MIN_VALUE : nums2[partitionB - 1];
            int minRightB = (partitionB == n) ? Integer.MAX_VALUE : nums2[partitionB];

            // Check if we have found the correct partitions
            if (maxLeftA <= minRightB && maxLeftB <= minRightA) {
                // If the total length of both arrays is even
                if ((m + n) % 2 == 0) {
                    return (
                        (Math.max(maxLeftA, maxLeftB) + // Max of left side
                        Math.min(minRightA, minRightB)) / // Min of right side
                        2.0 // Average them for median
                    );
                } else {
                    // If the total length is odd, return the max of the left side
                    return Math.max(maxLeftA, maxLeftB);
                }
            } else if (maxLeftA > minRightB) {
                // Move towards the left in nums1
                right = partitionA - 1;
            } else {
                // Move towards the right in nums1
                left = partitionA + 1;
            }
        }
        // If we exit the loop without finding the median, return 0.0 (should not happen)
        return 0.0;
    }
}
